### PR TITLE
Feature/simplify plugin factory plugin method

### PR DESCRIPTION
**Pull request by andsel**
This PR simplify the method `plugin` used to instantiate Java pipeline plugins, extracting common parts.

Mainly the changes regards the rewrite of a big `if else if ...` dispatching construct to use inheritance and map lookup. The code in single branches has been encapsulated into the subclasses of `AbstractPluginCreator`, where most of the common part is in the abstract class while the subclasses contains only the behavior specializations. A map name `pluginCreatorsRegistry`, constructed at `init` time in the `PluginFactoryExt`, acts as dispatcher to the various plugin's creator subclasses.

All 3 static classes present in previous `PluginFactoryExt` has been spoon off into self contained classes in the custom package `org.logstash.plugins.factory`.

### CODE REVIEW

File: logstash-core/src/main/java/org/logstash/plugins/factory/ExecutionContext.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import org.logstash.common.DLQWriterAdapter;
+import org.logstash.common.NullDeadLetterQueueWriter;
+import org.logstash.execution.ExecutionContextExt;
+import org.logstash.instrument.metrics.AbstractNamespacedMetricExt;
+import org.logstash.plugins.ContextImpl;
+import org.logstash.plugins.NamespacedMetricImpl;
+import org.logstash.plugins.PluginLookup;
+
+@JRubyClass(name = "ExecutionContextFactory")
+public final class ExecutionContext extends RubyBasicObject {
```
Comment (kares) [inline logstash-core/src/main/java/org/logstash/plugins/factory/ExecutionContext.java]: maybe naming Ruby/Java joined parts should follow some conventions consistently.
so in Ruby this is a `ExecutionContextFactory` but in Java its missing the `Factory` suffix, why?

File: logstash-core/src/main/java/org/logstash/plugins/factory/Metrics.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+import org.logstash.RubyUtil;
+import org.logstash.instrument.metrics.AbstractMetricExt;
+import org.logstash.instrument.metrics.AbstractNamespacedMetricExt;
+import org.logstash.instrument.metrics.MetricKeys;
+import org.logstash.instrument.metrics.NullMetricExt;
+
+import java.util.Arrays;
+
+@JRubyClass(name = "PluginMetricFactory")
+public final class Metrics extends RubyBasicObject {
```
Comment (kares) [inline logstash-core/src/main/java/org/logstash/plugins/factory/Metrics.java]: why are the Ruby/Java names so different, does it make sense?

File: logstash-core/src/main/java/org/logstash/plugins/factory/PluginsFactoryExt.java [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+import org.logstash.execution.ExecutionContextExt;
+import org.logstash.instrument.metrics.AbstractMetricExt;
+import org.logstash.instrument.metrics.AbstractNamespacedMetricExt;
+import org.logstash.instrument.metrics.MetricKeys;
+import org.logstash.plugins.PluginLookup;
+
+import java.util.*;
+
+@JRubyClass(name = "PluginFactory")
+public final class PluginsFactoryExt extends RubyBasicObject
```
Comment (kares) [inline logstash-core/src/main/java/org/logstash/plugins/factory/PluginsFactoryExt.java]: the convention seems to be adding an Ext suffix - thus `PluginFactoryExt` (wout the 's')

File: logstash-core/src/main/java/org/logstash/plugins/factory/ExecutionContext.java (hunk 1 above)
Comment (andsel) [reply logstash-core/src/main/java/org/logstash/plugins/factory/ExecutionContext.java]: These was inner classes, I kept the name. To be adherent to convention used it should be named `ExecutionContextExt`

File: logstash-core/src/main/java/org/logstash/plugins/factory/Metrics.java (hunk 2 above)
Comment (andsel) [reply logstash-core/src/main/java/org/logstash/plugins/factory/Metrics.java]: Should be named PluginMetricFactoryExt

**Comment by andsel** [comment]
Jenkins test this please

**Comment by andsel** [comment]
Jenkins test this please

File: logstash-core/src/main/java/org/logstash/plugins/factory/Metrics.java (hunk 2 above)
Comment (andsel) [reply logstash-core/src/main/java/org/logstash/plugins/factory/Metrics.java]: Done

File: logstash-core/src/main/java/org/logstash/plugins/factory/PluginsFactoryExt.java (hunk 3 above)
Comment (andsel) [reply logstash-core/src/main/java/org/logstash/plugins/factory/PluginsFactoryExt.java]: Done

File: logstash-core/src/main/java/org/logstash/plugins/factory/ExecutionContext.java (hunk 1 above)
Comment (andsel) [reply logstash-core/src/main/java/org/logstash/plugins/factory/ExecutionContext.java]: Renamed to `ExecutionContextExt`

File: logstash-core/src/main/java/org/logstash/plugins/factory/ExecutionContext.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    public ExecutionContext initialize(final ThreadContext context,
+                                       final IRubyObject agent, final IRubyObject pipeline, final IRubyObject dlqWriter) {
+        this.agent = agent;
+        this.pipeline = pipeline;
+        this.dlqWriter = dlqWriter;
+        return this;
+    }
+
+    @JRubyMethod
+    public ExecutionContextExt create(final ThreadContext context, final IRubyObject id,
```
Comment (kares) [inline logstash-core/src/main/java/org/logstash/plugins/factory/ExecutionContext.java]: :thinking: a bit confusing: `ExecutionContext` class is a factory for the `ExcutionContextExt` type ...
is it justified that we're not following the `#{RubyName}Ext` convention here?

File: logstash-core/src/main/java/org/logstash/plugins/factory/PluginFactoryExt.java [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+                        new IRubyObject[]{
+                                klass, typeScopedMetric, executionCntx,
+                                OutputStrategyExt.OutputStrategyRegistryExt.instance(context, null),
+                                rubyArgs
+                        }
+                );
+            } else if (type == PluginLookup.PluginType.FILTER) {
+                return filterDelegator(
+                        context, null,
+                        new IRubyObject[]{
```
Comment (kares) [inline logstash-core/src/main/java/org/logstash/plugins/factory/PluginFactoryExt.java]: :nail_care: would just change these to receive `IRubyObject...` as last arg (not much important)

File: logstash-core/src/main/java/org/logstash/plugins/factory/PluginMetricsFactory.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+import org.logstash.RubyUtil;
+import org.logstash.instrument.metrics.AbstractMetricExt;
+import org.logstash.instrument.metrics.AbstractNamespacedMetricExt;
+import org.logstash.instrument.metrics.MetricKeys;
+import org.logstash.instrument.metrics.NullMetricExt;
+
+import java.util.Arrays;
+
+@JRubyClass(name = "PluginMetricsFactory")
+public final class PluginMetricsFactory extends RubyBasicObject {
```
Comment (kares) [inline logstash-core/src/main/java/org/logstash/plugins/factory/PluginMetricsFactory.java]: personally, like that its named the same as the Ruby type but the codebase seems to use the `Ext` suffix convention

**Pull request closed without merging**